### ISSUE TITLE
Choose between single vs multiple miners per node architecture

### ISSUE THREAD

**Issue by malroc** [turinginc/credo-core-node#157]
In the current architecture, it is assumed that we have a single miner account per node. This has some unobvious consequences, particularly this means that we don't allow mobile users to participate in mining as they will be communicating to a single server.
Technically, we can allow nodes to maintain multiple miner accounts, which will require some refactoring, but if we want to allow mobile users to mine, it is worth doing sooner rather than later (it will be much harder to migrate to a different approach after launching the mainnet).
So basically the question is: wether we want to go with single vs multiple miners per node?
The great advantage of allowing multiple miners per node (i.e. allowing mobile users to mine) is that it almost eliminates the entry threshold for mining: any user with a mobile device and some Credo will be able to mine. However this also has its downside (aside from the fact that it requires additional work): it will be a huge step towards centralization, as we may expect that the number of mobile miners will be much more than the number of desktop ones, and all of them will be communicating to a single server/node.
@macgregordennis what are your thoughts on the above?

**Comment by macgregordennis** [comment]
> This has some unobvious consequences, particularly this means that we don't allow mobile users to participate in mining as they will be communicating to a single server.

@malroc It's not clear to me why mobile users wouldn't be able to mine with a single miner per node architecture. I think your comments assume that an app will be a client interfacing with some hosted server. However, we may be able to embed the CCN release within a mobile app, like we do with the native desktop app, or implement a mobile app client.

> what are your thoughts on the above?

My general thought is that I lean towards a single-miner per node architecture for the following reasons:

1. Closer resemblance to how we'd like most users to participate in mining (i.e. using a single account rather than multiple). Some elements of the our PoS algorithm's logic, such as a progressively dampened stake weight, will not apply if users mine with multiple mining accounts.
2. Increased barrier to mining centralization. If it's easy to create and operate multiple mining accounts, more users will likely do this.
3. Simplified logic on the back-end.
4. Simplified UI/UX for the wallet and other clients.

Basically, I think the singleton architecture is simpler, more decentralized, better conforms with ideal mining behavior, and should be mobile-compatible.

**Comment by malroc** [comment]
> @malroc It's not clear to me why mobile users wouldn't be able to mine with a single miner per node architecture. I think your comments assume that an app will be a client interfacing with some hosted server. However, we may be able to embed the CCN release within a mobile app, like we do with the native desktop app, or implement a mobile app client.

@macgregordennis first of all, we won’t be able to embed CCN into iOS app at least (not sure about Android) as Elixir is not supported as an iOS development language. So if we want mobile app to act as a node, we’ll have to reimplement node in some supported language.
Second, even if we do that, mining (which is a process that permanently consumes traffic and CPU time in the background) is an obvious battarey killer, hence I suppose an app like this will be rejected by the stores.
So I think that the only viable option to allow mining for mobile users is via client-server communication, which requires allowing multiple miner accounts per node. I am ok with staying with the current approach, but I just want to emphasize that in this case we make mobile mining impossible.

**Comment by macgregordennis** [comment]
@malroc I believe we should **not** distribute technology that we don't want people to use. Given we don't want people to mine with multiple accounts, since it's harder to apply incentive mechanisms to an actor operating multiple mining accounts and progressively dampen their mining power, we shouldn't release a version with that capability.

If we do use a client-server implementation for mobile users (which I'm still not convinced is necessarily the best approach given the alternatives of a native or embedded node with partial functionality), we may do so via a closed sourced fork. Perhaps you, @nishantj and I can this further on a call sometime soon.

**Comment by malroc** [comment]
@macgregordennis sounds reasonable. Re mobile apps with partial node functionality: right, that shouldn't be as complex as implementing the full-featured node (with mining etc), especially given that we already have the protocol, so we probably should go this way.
